homeassistant/components/feedreader/coordinator.py

```diff
@@ -42,72 +46,94 @@ def __init__(
-        self._feed = await self.hass.async_add_executor_job(self._fetch_feed)
+        def _parse_feed() -> feedparser.FeedParserDict:
+            return feedparser.parse(
+                self.url,
+                etag=None if not self._feed else self._feed.get("etag"),
+                modified=None if not self._feed else self._feed.get("modified"),
+            )
+
+        feed = await self.hass.async_add_executor_job(_parse_feed)
 
-        if not self._feed:
-            raise UpdateFailed(f"Error fetching feed data from {self._url}")
+        if not feed:
+            raise UpdateFailed(f"Error fetching feed data from {self.url}")
 
         # The 'bozo' flag really only indicates that there was an issue
         # during the initial parsing of the XML, but it doesn't indicate
         # whether this is an unrecoverable error. In this case the
         # feedparser lib is trying a less strict parsing approach.
         # If an error is detected here, log warning message but continue
         # processing the feed entries if present.
-        if self._feed.bozo != 0:
-            if isinstance(self._feed.bozo_exception, URLError):
+        if feed.bozo != 0:
+            if isinstance(feed.bozo_exception, URLError):
                 raise UpdateFailed(
-                    f"Error fetching feed data from {self._url}: {self._feed.bozo_exception}"
+                    f"Error fetching feed data from {self.url} : {feed.bozo_exception}"
                 )
 
             # no connection issue, but parsing issue
             _LOGGER.warning(
                 "Possible issue parsing feed %s: %s",
-                self._url,
-                self._feed.bozo_exception,
+                self.url,
+                feed.bozo_exception,
             )
+        return feed
+
+    async def async_setup(self) -> None:
+        """Set up the feed manager."""
+        feed = await self._async_fetch_feed()
+        self.logger.debug("Feed data fetched from %s : %s", self.url, feed["feed"])
+        self.feed_author = feed["feed"].get("author")
+        self.feed_version = feedparser.api.SUPPORTED_VERSIONS.get(feed["version"])
+        self._feed = feed
+
+    async def _async_update_data(self) -> list[feedparser.FeedParserDict] | None:
+        """Update the feed and publish new entries to the event bus."""
+        assert self._feed is not None
+        # _last_entry_timestamp is not set during async_setup, but we have already
+        # fetched data, so we can use them, instead of fetch again
+        if self._last_entry_timestamp:
+            self._feed = await self._async_fetch_feed()
+
         # Using etag and modified, if there's no new data available,
         # the entries list will be empty
         _LOGGER.debug(
             "%s entri(es) available in feed %s",
             len(self._feed.entries),
-            self._url,
+            self.url,
         )
-        if not self._feed.entries:
+        if not isinstance(self._feed.entries, list):
```

Comment:
Why did we change this check?

Reply:
to ensure we return None or a list, but not `Any`

Reply:
Ok. Is it just missing typing in the library or can that actually happen? If it's the former, maybe just ignore the type instead?

Reply:
it is about missing typing, so yes your suggestion to ignore the typing would be the better solution - will add this to a follow up PR 👍

Reply:
#122327